### PR TITLE
Add 'index_col' parameter to DataFrame.to_spark

### PR DESCRIPTION
**Pull request by HyukjinKwon**
This PR adds `index_col` parameter to `DataFrame.to_spark`. This PR relates to https://github.com/databricks/koalas/pull/901.

After this PR, we can do a roundtrip in Koalas and Spark DataFrame easily.

```python
>>> import databricks.koalas as ks
>>> kdf = ks.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6], 'c': [7, 8, 9]})
>>> sdf = kdf.to_spark(index_col="index").filter("a == 2")  # PySpark API
>>> kdf = sdf.to_koalas(index_col="index")
>>> kdf
       a  b  c
index
1      2  5  8
```

### CODE REVIEW

File: databricks/koalas/internal.py [hunk 1]
```diff
@@ -599,34 +605,30 @@ def column_index_names(self) -> Optional[List[str]]:
         """ Return names of the index levels. """
         return self._column_index_names
 
+    def _data_column_names(self):
+        index_columns = set(self.index_columns)
+        data_columns = []
+        for i, (column, idx) in enumerate(zip(self._data_columns, self.column_index)):
+            assert column not in index_columns
```
Comment (HyukjinKwon) [inline databricks/koalas/internal.py]: @ueshin, mind if I ask to double check if this is correct?

File: databricks/koalas/frame.py [hunk 2]
```diff
@@ -2742,7 +2741,7 @@ def duplicated(self, subset=None, keep='first'):
                                              index_map=[(index_column,
                                                          self._internal.index_names[0])])))
 
-    def to_koalas(self, index_col=None):
+    def to_koalas(self, index_col: Optional[Union[str, List]] = None):
```
Comment (ueshin) [inline databricks/koalas/frame.py]: `Optional[Union[str, List[str]]`?

File: databricks/koalas/frame.py [hunk 3]
```diff
@@ -3064,18 +3063,90 @@ def to_spark_io(self, path: Optional[str] = None, format: Optional[str] = None,
         self.to_spark().write.save(
             path=path, format=format, mode=mode, partitionBy=partition_cols, options=options)
 
-    def to_spark(self):
+    def to_spark(self, index_col: Optional[Union[str, List]] = None):
```
Comment (ueshin) [inline databricks/koalas/frame.py]: ditto.

File: databricks/koalas/frame.py [hunk 4]
```diff
@@ -3064,18 +3063,90 @@ def to_spark_io(self, path: Optional[str] = None, format: Optional[str] = None,
...
+        index_1 index_2
+        0       1        4  7
+        1       2        5  8
+        2       3        6  9
+        """
+        if index_col is None:
+            return self._internal.spark_df
+        else:
+            if isinstance(index_col, str):
+                index_col = [index_col]
```
Comment (ueshin) [inline databricks/koalas/frame.py]: We should check the length of `index_col` to match the number of index?

Comment (ueshin) [reply databricks/koalas/frame.py]: Also we should check the names are not duplicated with data column names.

File: databricks/koalas/internal.py [hunk 5]
```diff
@@ -398,11 +403,12 @@ def __init__(self, sdf: spark.DataFrame,
         assert isinstance(sdf, spark.DataFrame)
         if index_map is None:
             # Here is when Koalas DataFrame is created directly from Spark DataFrame.
-            assert "__index_level_0__" not in sdf.schema.names, \
-                "Default index column should not appear in columns of the Spark DataFrame"
+            assert any(SPARK_INDEX_NAME_PATTERN.match(name) for name in sdf.schema.names), \
```
Comment (ueshin) [inline databricks/koalas/internal.py]: `not any(SPARK_INDEX_NAME_PATTERN.match(name) for name in sdf.schema.names)` or `all(not SPARK_INDEX_NAME_PATTERN.match(name) for name in sdf.schema.names)`?

File: databricks/koalas/internal.py (hunk 1 above)
Comment (ueshin) [reply databricks/koalas/internal.py]: Actually we do have such a case. We can't raise an error.

Comment (HyukjinKwon) [reply databricks/koalas/internal.py]: Hmmmm .. okie let me revert it back for now.

File: databricks/koalas/internal.py [hunk 6]
```diff
@@ -36,6 +36,11 @@
 from databricks.koalas.utils import column_index_level, default_session, lazy_property, scol_for
 
 
+# A function to turn given numbers to Spark columns that represent Koalas index.
+SPARK_INDEX_NAME_FORMAT = "__index_level_{}__".format
+# A pattern to check if the name of a Spark column is a Koalas index name or not.
+SPARK_INDEX_NAME_PATTERN = re.compile(r"__index_level_[0-9]+__")
```
Comment (HyukjinKwon) [inline databricks/koalas/internal.py]: These changes are not related actually ..

**Comment by ueshin** [comment]
Thanks! merging.

**Comment by HyukjinKwon** [comment]
Thanks!!
